cpp/src_prims/matrix/batched_matrix.h

```diff
@@ -0,0 +1,465 @@
+#pragma once
+
+#include <functional>
+#include <stdexcept>
+#include <tuple>
+#include <unordered_map>
+#include <vector>
+
+#include <linalg/binary_op.h>
+#include <linalg/cublas_wrappers.h>
+#include <memory>
+
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+
+namespace MLCommon {
+namespace Matrix {
+
+std::shared_ptr<double*> init(double* A, std::pair<int, int> shape,
+                              int num_batches, bool gpu = true);
+
+__global__ void identity_matrix_kernel(double** I, int r);
+
+template <typename T>
+void cudaFreeT(T* ptr) {
+  CUDA_CHECK(cudaFree(ptr));
+}
+
+void BMM_Allocate(std::pair<int, int> shape, int num_batches, double*& A_dense,
+                  double**& A_array, bool setZero) {
+  int m = shape.first;
+  int n = shape.second;
+
+  allocate(A_dense, m * n * num_batches, setZero);
+  allocate(A_array, num_batches);
+
+  // fill array of pointers to each batch matrix.
+  auto counting = thrust::make_counting_iterator(0);
+  thrust::for_each(counting, counting + num_batches, [=] __device__(int bid) {
+    A_array[bid] = &(A_dense[bid * m * n]);
+  });
+}
+
+// https://stackoverflow.com/questions/32685540/why-cant-i-compile-an-unordered-map-with-a-pair-as-key
+struct pair_hash {
+  template <class T1, class T2>
+  std::size_t operator()(const std::pair<T1, T2>& p) const {
+    auto h1 = std::hash<T1>{}(p.first);
+    auto h2 = std::hash<T2>{}(p.second);
+
+    // Mainly for demonstration purposes, i.e. works but is overly simple
+    // In the real world, use sth. like boost.hash_combine
+    return 100 * h1 + h2;
+  }
+};
+
+struct BatchedMatrixMemory {
+  BatchedMatrixMemory(std::pair<int, int> shape, int num_batches,
+                      bool setZero) {
+    BMM_Allocate(shape, num_batches, A_dense, A_array, setZero);
+    // std::cout << "Allocate (" << shape.first << "," << shape.second << "):" << A_array << "\n";
+    in_use = true;
+  }
+
+  double* A_dense;
+  double** A_array;
+  bool in_use;
+};
+
+class BatchedMatrixMemoryPool {
+ public:
+  BatchedMatrixMemoryPool(int num_batches) : m_num_batches(num_batches) {
+    // std::cout << "Memory Pool Init\n";
+    CUBLAS_CHECK(cublasCreate(&m_handle));
```

Comment:
inside ml-prims, we don't create any of these handles. Please update the interface to accept it as an argument, instead.